README.md
# DoSK
Docker Starter Kit - стартовое окружение php-разработчика основанное на `docker`.

## Версионирование образов

Папки без указания версий - текущая актуальная версия для ведения разработки. 
Если она устарела, или нужна специфическая версия - заводим отдельную папку для образа, с указанием версии. 
Например понадобился php-fpm версии 8.0, а текущий 7.4. 
Создаём рядом папку php-fpm-8.0 разворачиваем конкретный образ в нём. 

## Описание файлов и каталогов

* [`/app`](./app) - каталог с приложением, либо подпапки
* [`/data`](./data) - каталог с данными от БД, эластика и т.п.
* [`/docker`](./docker) - каталог с файлами настройки образов
* [`/scripts`](./scripts) - каталог с вспомогательными скриптами 
* [`.env.example`](.env.example) - пример файла переменных окружения
* [`docker-compose.yml`](docker-compose.yml) - основной файл сборки контейнеров

## Первоначальная настройка

### Для Linux

#### Файл локального окружения

Просто копируем заготовку в `.env` файл.
```bash
$ cp ./.env.example ./.env 
```
В `.env` указываем необходимые переменные окружения.

#### Пользователь и группа, для запуска web-сервера
Перед началом использования Linux-пользователям надо добавить группу и пользователя, а так же включить своего пользователя в эту группу.
Сделать это можно так:
```bash
$ sudo groupadd --gid 1002 --system containers_group
$ sudo useradd --uid 1002 --system --shell /bin/false --gid 1002 containers_user
$ sudo usermod -aG containers_group $USER
```
Этот пользователь и группа будут использоваться для `nginx` и `php-fpm` в контейнерах, для того чтобы они могли общаться через 
unix socket. Не уверен что выполненно полностью корректно, с точки зрения безопасности, но для локального окружения сойдёт. 

Строки `containers_user` и `containers_group` можно заменить по вкусу, они роли не играют, но я бы советовал оставить такими же - для единообразия. 
`UID` и `GID`, указанные как `1002`, лучше уточнить в `docker-compose.yml`, их при необходимости тоже можно заменить.

#### Настройки ядра
##### Для Elasticsearch
Для нормальной работы ElasticSearch желательно поправить следующие параметры ядра `/etc/sysctl.d/99-sysctl.conf`:
```bash
vm.max_map_count=262144
# пороговый процент свободной оперативной памяти, для использования файла подкачки
vm.swappiness=10
```
Для этого нужно добавить вышеуказанную строку в файл /etc/sysctl.conf или в последний файл в каталоге /etc/sysctl.d. Для вступления изменений в силу нужно перезагрузиться, или применить вручную:
```bash
$ echo 262144 > /proc/sys/vm/max_map_count
``` 

### Для Windows
` to be continue...`

## Использование

### hosts файл
Необходимо добавить доменные имена в файл `hosts`
```bash
$ sudo echo "127.0.0.1  app.lc" >> /etc/hosts 
```

### Скрипты запуска
Размещены в каталоге `./scripts`. 
Нижеописанные скрипты удобно использовать для отладки образов, или просто для локального запуска:
* `init.sh`  служит для инициализации (проверено на Ubuntu 18.04)
* `rebuild.sh` служит хелпером для пересборки конкретного контейнера
* `shell.sh` запускает шелл для пользователя в контейнере `app`

Скрипты размещённые в папках с образами:
* `build.sh` нужен для сбора образа
* `run.sh` нужен для запуска контейнера в ручную

Скрипты не нужны при запуске с помощью Docker compose.

### Использование docker-compose
Можно использовать `docker-compose.yml` из консоли, с командой `docker-compose`:
```bash
$ docker-compose -f docker-compose.yml build
$ docker-compose -f docker-compose.yml up
$ docker-compose -f docker-compose.yml down
```

### Запуск окружения из PHPStorm
В PHPStorm есть функционал, для удобного запуска контейнеров описанных в docker-compose.yml файлах. 
Для этого нужно добавить конфигурацию Запуска/Отладки: в главном меню нажимаем: `Run -> Edit Configurations...` в открывшемся окне, вверху слева жмём на `+` и выбираем `Docker -> Docker-compose`. 
Даём название конфигурации, убеждаемся что PHPStorm нашёл запущенный на компьютере Docker и отобразил его в графе `Server`. Теперь в графе `Compose file(s)` жмём на пиктограмму папки, затем на `+` и выбираем файл `docker-compose.yml`. Теперь при запуске добавленной конфигурации создадутся все необходимые образы, а из них контейнеры, а после их запуска, всё необходимое окружение будет готово.  

## Внедрение в проект
Скопировать в корень проекта папку docker, со всем содержимым и файл `docker-compose.yml`.

Следует проверить сервисы описанные в `docker-compose.yml` и удалить не нужные

## Подключение к базе данных
Создайте в папке пользователя файл `.my.cnf`
```bash
$ touch ~/.my.cnf
```
 и добавьте туда следующее содержимое:
```bash
[client]
host=127.0.0.1
port=$MYSQL_PORT
user=root
password=$MYSQL_ROOT_PASSWORD
```
теперь можете подключиться к БД из консоли
```bash
$ sudo apt-get install mysql-client
$ mysql
```
### Восстановление из дампа
Устанавливаем необходимые пакеты
```bash
$ sudo apt-get install wget unzip
```
затем скачиваем файл дампа по предоставленной ссылке, распаковываем, заливаем в БД и удаляем
```bash
$ wget $MYSQL_DUMP_URL -O /tmp/dump.zip
$ unzip -p /tmp/dump.zip | mysql --database=$MYSQL_DATABASE
$ rm /tmp/dump.zip
```

## Планы на будущее

* добавить в контейнер `application`
    * поддержку node.js
    * поддержку `crontab`
* изолировать контейнер `php-cli`
* добавить в php-fpm
    * AMQP, rabbitmq
    * memcached 
* `to be continue...`

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

CHANGELOG.md
# Changelog

All notable changes to `patrikap/dosk` will be documented in this file.

## 1.0.0 - 2020-04-03
- Initial release

## 1.0.1 - 2020-04-13
- Add all containers

## 1.0.2 - 2020-04-14
- Add support `pgsql` and `pdo_pgsql`
- Add git support in `application`

scripts/init.sh
#!/usr/bin/env bash
# переходим науровень выше
cd ../
### READ .ENV
read_var() {
    VAR=$(grep $1 $2 | xargs)
    IFS="=" read -ra VAR <<< "$VAR"
    echo ${VAR[1]}
}
### CONFIG
GID=$(read_var GID .env)
UID=$(read_var UID .env)
CONTAINER_GROUP=containers_group
CONTAINER_USER=containers_user

# copy .env file
cp .env.example .env

sudo groupadd --gid ${GID} --system ${CONTAINER_GROUP}
sudo useradd --uid ${UID} --system --shell /bin/false --gid ${GID} ${CONTAINER_USER}
sudo usermod -aG ${CONTAINER_GROUP} ${USER}

sudo echo "vm.max_map_count=262144" >> /etc/sysctl.d/99-sysctl.conf
sudo echo "vm.swappiness=10" >> /etc/sysctl.d/99-sysctl.conf
sudo echo 262144 > /proc/sys/vm/max_map_count

# переименовать/переместить папку по названию проекта
# удалит папку с гитом
#rm -rf ./git

# возвращаемся в папку со скриптами
cd ./scripts
